kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs

```diff
@@ -916,26 +917,80 @@ impl<'tcx> GotocCtx<'tcx> {
+                1,
+                "Expected a single-variant ADT. Found {:?}",
+                typ
+            );
+            let fields = &adt_def.variants().get(VariantIdx::from_u32(0)).unwrap().fields;
+            let mut non_zsts = fields
+                .iter()
+                .filter(|field| !self.layout_of(field.ty(self.tcx, adt_substs)).is_zst());
+            let non_zst = non_zsts.next().expect("Expected one non-zst field.");
+            assert!(non_zsts.next().is_none(), "Expected only one non-zst field.");
+            Some((non_zst.name.to_string(), non_zst.ty(self.tcx, adt_substs)))
+        } else {
+            None
+        }
+    }
+
+    /// Extract the fat pointer out of the self receiver type.
+    ///
+    /// As of Jul 2022:
+    /// P = &Self | &mut Self | Box<Self> | Rc<Self> | Arc<Self>
+    /// S = P | Pin<P>
+    ///
+    /// See https://doc.rust-lang.org/reference/items/traits.html#object-safety for more details.
+    fn receiver_fat_ptr(&self, arg_ty: Ty<'tcx>) -> Ty<'tcx> {
+        trace!(mir_ty=?arg_ty, "receiver_fat_ptr");
+        let mut typ = arg_ty;
+        // Traverse the layout following the non-zero field until we find a fat pointer.
+        while let Some((_, new_typ)) = self.get_non_zst_field(typ) {
+            typ = new_typ;
+        }
+
+        assert!(self.is_vtable_fat_pointer(typ));
+        trace!(?typ, "receiver_fat_ptr");
+        typ
+    }
+
     /// Generate code for a trait function declaration.
     ///
-    /// Dynamic function calls first parameter is the fat-pointer representing self.
-    /// For closures, the type of the first argument is dyn T not &dyn T.
-    pub fn codegen_dynamic_function_sig(&mut self, sig: PolyFnSig<'tcx>) -> Type {
+    /// Dynamic function calls first parameter is self which must be one of the following:
+    ///
+    /// As of Jul 2022:
+    /// P = &Self | &mut Self | Box<Self> | Rc<Self> | Arc<Self>
+    /// S = P | Pin<P>
+    ///
+    /// See https://doc.rust-lang.org/reference/items/traits.html#object-safety for more details.
+    fn codegen_dynamic_function_sig(&mut self, sig: PolyFnSig<'tcx>) -> Type {
         let sig = self.monomorphize(sig);
         let sig = self.tcx.normalize_erasing_late_bound_regions(ty::ParamEnv::reveal_all(), sig);
-
         let mut is_first = true;
         let params = sig
             .inputs()
             .iter()
             .filter_map(|arg_type| {
                 if is_first {
-                    // This should &dyn T or dyn T (for closures).
                     is_first = false;
-                    let first_ty = pointee_type(*arg_type).unwrap_or(*arg_type);
                     debug!(self_type=?arg_type, fn_signature=?sig, "codegen_dynamic_function_sig");
-                    assert!(first_ty.is_trait(), "Expected self type to be a trait");
-                    Some(self.codegen_trait_data_pointer(first_ty))
+                    let self_ty = if arg_type.is_ref() {
```

Comment:
What might be helpful here is a sketch of what the memory layouts look like for these cases